### DISCUSSION TITLE
SFRA - Some of the middleware events are not firing

### DISCUSSION THREAD

**Discussion by amitkathpal123** [SalesforceCommerceCloud/sfcc-ci#415] [Team Posts]
Hi All,

I am testing to fire MW events (step, start, beforeComplete) by calling a MW function.
 - Route:beforeComplete is firing correctly
 - MW function is also called successfully
 - Route:Step and Route:Start are not firing

Can you please check below code and point me to the mistake or if anyone has sample code where all events are working will be helpful.

// Controller Function
server.get('EventEmitters',  testmw.testMWFunc, server.middleware.https, function(req, res, next){
	
	//this.on('route:Redirect', function (req, res) {
	//	res.setViewData({valueRedirect:'This is the text from event Route:Redirect'});
	//})

	this.on('route:Start', function (req, res) {
		res.setViewData({valueStart:'This is the text from event Route:Start'});
	})
	
	this.on('route:Step', function (req, res) {
		res.setViewData({valueStep:'This is the text from event Route:Step'});
	})
	
	this.on('route:BeforeComplete', function (req, res) {
		res.setViewData({valueBeforeComplete:'This is the text from event Route:BeforeComplete'});
	})
	
	res.render("eeTemplate",{value:'This text is from Main Function'})
	next();
});

Thanks in advance

**Comment by zlatinz** [comment]
Hi @amitkathpal123 ,

In your code you always replace the viewData and never keep the value of attributes set by previous event handler calls. This is why in the end you see only single attribute, which is set from the last event handler executed.

You need to follow the following code pattern to get the result you are probably expecting:
```javascript
this.on('route:Start', function (req, res) {
        var viewData = res.getViewData();
        viewData.valueStart = 'This is the text from event Route:Start';
	res.setViewData(viewData);
});
```

Hope this helps,
Zlatin

**Comment by amitkathpal123** [comment]
Thanks @zlatinz . I tried but it dint make any difference.
Just read, setViewData function merges all the values set after each MW function.

**Comment by zlatinz** [comment]
Just checked, and you are right, setViewData uses assign.js to do a copy of all viewData object properties.

Have you tried this?

```javascript
// Controller Function
var eventEmitters = server.get('EventEmitters', 
	testmw.testMWFunc, 
	server.middleware.https, 
	function(req, res, next) {
		res.render("eeTemplate",{value:'This text is from Main Function'})
		next();
	}
);

eventEmitters.on('route:Start', function (req, res) {
	res.setViewData({valueStart:'This is the text from event Route:Start'});
});
eventEmitters.on('route:Step', function (req, res) {
	res.setViewData({valueStep:'This is the text from event Route:Step'});
});
eventEmitters.on('route:BeforeComplete', function (req, res) {
	res.setViewData({valueBeforeComplete:'This is the text from event Route:BeforeComplete'});
});
```

**Comment by amitkathpal123** [comment]
Thanks @zlatinz . This worked but wondering what difference it made.

**Comment by zlatinz** [comment]
Hi @amitkathpal123 ,
The difference is that when you put the event handlers in the code of the middleware function, the event handlers are enabled after the middleware function is executed, so some events like route:Start has already been fired, and you cannot capture them.
When you do the event handlers outside, they are activated the moment route is defined, so you can capture.

```javascript
// Controller Function
var eventEmitters = server.get('EventEmitters', 
	testmw.testMWFunc, 
	server.middleware.https, 
	function(req, res, next) {
		this.on('route:Start', function (req, res) {
			res.setViewData({valueStartInside:'This is the text from event Route:Start'});
		});
		res.render("eeTemplate",{value:'This text is from Main Function'})
		next();
	}
);

eventEmitters.on('route:Start', function (req, res) {
	res.setViewData({valueStartOutside:'This is the text from event Route:Start'});
});
```
The sequence of execution is (simplifying things a bit):
1. **server.get('EventEmitters', ...** - definition of the route
2. **eventEmitters.on('route:Start', ...** - definition of an outside event handler
3. request comes from customer that hits **yourController-EventEmitters** route
4. **route:Start** is emitted
5. **route:Step** is emitted for **server.middleware.get** middleware function (added to the route middleware chain from server.get call)
6.  **server.middleware.get** is executed
7. **route:Step** is emitted for **testmw.testMWFunc** middleware function 
8. **testmw.testMWFunc** is executed
9. **route:Step** is emitted for **server.middleware.https** middleware function 
10. **server.middleware.https** is executed
11. **route:Step** is emitted for **function(req, res, next)** defined in EventEmitters route
12. **function(req, res, next)** defined in EventEmitters route is executed
13. As part of the execution of 12. **this.on('route:Start',** is executed and another event handler is attached to the route:Start event, which was emitted in step 4
14. **route:BeforeComplete** is emitted, as there are no more functions in the middleware chain of EventEmitters route.
15. **route:Complete** is emitted and the corresponding templates are rendered

**Comment by amitkathpal123** [comment]
@zlatinz . Thank you. This clarifies and helps.
